Forms: add tracking pixel to form submission emails

vianasw opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fixes FORMS-54

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Added tracking pixel to the form submission emails

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Spin a JN site with this branch.
* Create a simple form and publish it.
* Submit and go to your email inbox to check the form submission email.
* After opening the email, go to Tracks and check that the event `jetpack_forms_email_open` shows up (it might take a few minutes).

vianasw committed:
Add tracking pixel to email form submission

Add tracking pixel to the form submission emails. To comply with GDPR,
we're not tracking any user information, just the email open email and a
hash of the admin email.

vianasw committed:
changelog

vianasw committed:
Fix static analysis error

vianasw commented on projects/packages/forms/src/contact-form/class-contact-form.php:
```diff
@@ -1861,8 +1862,16 @@ public static function wrap_message_in_html_tags( $title, $body, $footer ) {
+		$event          = new Jetpack_Tracks_Event(
+			(object) array(
+				'_en' => 'jetpack_forms_email_open',
+				'_ui' => hash_hmac( 'md5', get_option( 'admin_email' ), JETPACK__VERSION ),
```
This hash here is just to have a unique, anonymized string for this field. The secret key used in the hash_hmac method is not that relevant (we're not going to validate it). So, I used JETPACK__VERSION, but I could have used any other constant.

simison replied:
Above could be a code comment too, for the next one looking at this?

simison commented on projects/packages/forms/src/contact-form/class-contact-form.php:
```diff
@@ -1861,8 +1862,16 @@ public static function wrap_message_in_html_tags( $title, $body, $footer ) {
+				'_ut' => 'anon',
+			)
+		);
+		$tracking_pixel = '<img src="' . $event->build_pixel_url() . '" />';
```
Can you add `alt="" width="1" height="1"` to ensure the tracking pixel image stays at 1px width/height even when it doesn't load due to DNS or some plugin blocking it from loading?

I fixed a similar issue in another email some time ago:

![image](https://github.com/user-attachments/assets/7209a213-51ca-4acd-adbf-b7f71af2e6f0)

`alt=""` avoids screen readers from seeing the image.

simison commented on projects/packages/forms/src/contact-form/class-contact-form.php:
```diff
@@ -1861,8 +1862,16 @@ public static function wrap_message_in_html_tags( $title, $body, $footer ) {
+		$style          = '';
+		$event          = new Jetpack_Tracks_Event(
+			(object) array(
+				'_en' => 'jetpack_forms_email_open',
```
Event name looks good and follows Tracks event naming guidelines 👍

simison reviewed (dismissed):
Would be good if you can enforce 1px width/height before merging, but otherwise looks good and tests well!

Pixel is in the email:
<img width="1653" alt="Screenshot 2025-05-27 at 12 57 47" src="https://github.com/user-attachments/assets/9df439a0-846d-40c4-96ac-b802892fb82e" />

And Tracks event came through:
<img width="1222" alt="Screenshot 2025-05-27 at 12 56 26" src="https://github.com/user-attachments/assets/61f1bca8-7d78-4699-a6ae-95ab96550e28" />

Should we add blogid to the event as well? Not a blocker for merging tho.

vianasw committed:
Address PR comments

I added widgth and height attributes to the img that contains the
tracking pixel to ensure the tracking pixel image stays at 1px
width/height even when it doesn't load due to DNS or some plugin
blocking it from loading.

I also added some clarifying comments regarding the use of the hash
function and the secret key.
